lib/davclient.h

```diff
@@ -0,0 +1,96 @@
+ * version 2.1 as published by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA
+ */
+
+#ifndef DAVCLIENT_H
+#define DAVCLIENT_H
+
+#include <QObject>
+#include <QSharedDataPointer>
+#include <QDateTime>
+#include <QHash>
+#include <QNetworkReply>
+
+#include "davexport.h"
+#include "davtypes.h"
+
+namespace DAV {
+class ClientPrivate;
+
+class DAV_EXPORT Client : public QObject
+{
+    Q_OBJECT
+
+public:
+    struct Reply
+    {
+        QString uri;
+        QNetworkReply::NetworkError networkError;
+        QString errorMessage;
+        QByteArray errorData;
+
+        Reply(const QString &path, QNetworkReply::NetworkError error,
+              const QString &message, const QByteArray &data)
+            : uri(path), networkError(error), errorMessage(message), errorData(data) {}
+        bool errorOccured() const {return networkError != QNetworkReply::NoError;}
+    };
+    
+    Client(const QString &serverAddress, const QString &rootPath,
+           QObject *parent = nullptr);
+    ~Client();
+
+    QString serverAddress() const;
+    QString rootPath() const;
+
+    bool ignoreSSLErrors() const;
+    void setIgnoreSSLErrors(bool ignore);
+    
+    void setAuthLogin(const QString &username, const QString &password);
+    void setAuthToken(const QString &token);
+
+    void requestUserPrincipalData();
+    QString userPrincipal() const;
+    QString userPrincipalMailto() const;
+    QString userPrincipalHome() const;
+
+    void requestCalendarList(const QString &path = QString());
+    QList<CalendarInfo> calendars() const;
+
+    void getICalEtags(const QString &path,
+                      const QDateTime &from, const QDateTime &to);
+    void getICalResources(const QString &path,
+                          const QDateTime &from, const QDateTime &to);
+    void getICalResources(const QString &path, const QStringList &uids);
+    void sendICalResource(const QString &path, const QString &data, const QString &etag = QString());
+    void deleteResource(const QString &path);
+
+signals:
+    void userPrincipalDataFinished(const Reply &reply);
+    void calendarListFinished(const Reply &reply);
+    void iCalEtagsFinished(const Reply &reply, const QHash<QString, QString> &etags);
+    void iCalResourcesFinished(const Reply &reply, const QList<ICalResource> &resources);
+    void sendICalFinished(const Reply &reply, const QString &etag);
+    void deleteFinished(const Reply &reply);
+
+private:
+    QSharedDataPointer<ClientPrivate> d;
```

Comment:
Maybe not that much point in QSharedDataPointer & QSharedData inheritance on private. The client shouldn't be copied or shared.